src/RestApi/StoreApi/Utilities/OrderController.php

```diff
@@ -97,6 +97,100 @@ public function sync_customer_data_with_order( \WC_Order $order ) {
 		};
 	}
 
+	/**
+	 * Final validation ran before payment is taken.
+	 *
+	 * By this point we have an order populated with customer data and items.
+	 *
+	 * @throws RouteException Exception if invalid data is detected.
+	 *
+	 * @param \WC_Order $order Order object.
+	 */
+	public function validate_order_before_payment( \WC_Order $order ) {
+		$coupons = $order->get_coupon_codes();
+
+		$coupon_errors = [];
+
+		foreach ( $coupons as $coupon_code ) {
+			$coupon = new \WC_Coupon( $coupon_code );
+
+			try {
+				$this->validate_coupon_email_restriction( $coupon, $order );
+			} catch ( \Exception $error ) {
+				$coupon_errors[ $coupon_code ] = $error->getMessage();
+			}
+			try {
+				$this->validate_coupon_usage_limit( $coupon, $order );
+			} catch ( \Exception $error ) {
+				$coupon_errors[ $coupon_code ] = $error->getMessage();
+			}
+		}
```

Comment:
I see that we're validating case by case here, why aren’t we using Core validation?

Reply:
The core functions make use of global `POST` so it's not suitable for us.

Reply:
so in a follow up next cycle, we should see how those extensions p2-pbIy4N-7d are doing it, ideally, it would be a couple of filters or actions here and there.